### ISSUE TITLE
Schnorr signatures

### ISSUE THREAD

**Issue by daurnimator** [openssl/openssl#8440]
For a project, I'm looking for an implementation of [Schnorr Signatures](https://en.wikipedia.org/wiki/Schnorr_signature) to store many signatures within a small constant size.

Is this something that would fit into OpenSSL?

**Comment by paulidale** [comment]
The policy is:

> All algorithms and protocols should be recognized by a national or international standards body. That is somewhat vague, but the important point is that we most of us are implementors, not cryptographers, and will defer judgement to experts.

From [here](https://www.openssl.org/blog/blog/2018/01/18/f2f-london/).

I haven't easily found a standard describing this algorithm.  [RFC 8235](https://tools.ietf.org/html/rfc8235) seems related but not quite the same.  Do you know of such a standard?

**Comment by daurnimator** [comment]
Hrm, the only other "standard" than RFC 8235 that I can find that uses schnorr signatures is a bitcoin related one: https://github.com/sipa/bips/blob/bip-schnorr/bip-schnorr.mediawiki

**Comment by mattcaswell** [comment]
Probably that doesn't count as coming from a "national or international standards body".

**Comment by InfoHunter** [comment]
The first time I encountered with this algorithm is in a book called "Cryptography Theory and Practice" (Third Edition).

**Comment by mattcaswell** [comment]
Closing this since there does not seem to be a suitable standard.

**Comment by michaelfolkson** [comment]
I was looking to see what other cryptography libraries support Schnorr signatures (or are planning to support Schnorr signatures) and I found this issue.

> Hrm, the only other "standard" than RFC 8235 that I can find that uses schnorr signatures is a bitcoin related one: https://github.com/sipa/bips/blob/bip-schnorr/bip-schnorr.mediawiki

Today the Bitcoin BIP (Bitcoin Improvement Proposal) for Schnorr signatures has been assigned a BIP number (340), merged into the BIPs repo, is active on the Bitcoin network and is effectively [finalized](https://github.com/bitcoin/bips/blob/master/bip-0340.mediawiki) (despite still being labeled draft in the BIPs repo).

> Probably that doesn't count as coming from a "national or international standards body".

I'd be interested in what criteria need to be met to be considered a "national or international standards body" if anyone can point to me some resources? That isn't to say I disagree with this statement, I am honestly unsure and can see arguments for and against this statement.

**Comment by GregoryTrzonkowski** [comment]
> (...) should be recognized by a national or international standards body. That is somewhat vague, but the important point is that we most of us are implementors, not cryptographers, and will defer judgement to experts.

Is the https://www.enisa.europa.eu/publications/algorithms-key-sizes-and-parameters-report/at_download/fullReport recommendation good enough as a recognition, perhaps? There is the quote:

> we [European Union Agency for Cybersecurity) recommend Schnorr signatures [313], which have stronger provable security properties than DSA [254, 283].

**Comment by mattcaswell** [comment]
> Is the https://www.enisa.europa.eu/publications/algorithms-key-sizes-and-parameters-report/at_download/fullReport recommendation good enough, perhaps? 

No. It needs to be an actual standard that we can follow as implementers detailing specifically how the scheme works and published by a "national or international standards body".

Note though that the requirement to have a standard only applies to code that is actually going to get merged into the OpenSSL codebase. With OpenSSL 3.0 we have the ability to plug-in 3rd party providers which can be used to supply a custom signature scheme. So there's no reason someone couldn't implement an external provider implementing Schnorr signatures and have it work with off-the-shelf OpenSSL 3.0.

**Comment by GregTonoski** [comment]
Thanks for the detailed response.

On a side note, there is the open source C implementation of Schnorr signatures: [https://github.com/metalicjames/BIP340/tree/master/C](https://github.com/metalicjames/BIP340/tree/master/C)

**Comment by rb-anssi** [comment]
Hi,

To be complete, Schnorr based signatures (both original SDSA based on "regular" Schnorr groups, and EC-SDSA based on elliptic curves groups) are standardized in ISO14888-3 (https://webstore.iec.ch/preview/info_isoiec14888-3%7Bed4.0%7Den.pdf). EC-SDSA comes in three variants (mainly for optimization of the signature size) and each one has a registered OID ([1.0.14888.3.0.11](http://oid-info.com/get/1.0.14888.3.0.11), [1.0.14888.3.0.13](http://oid-info.com/get/1.0.14888.3.0.13), [1.0.14888.3.0.12](http://oid-info.com/get/1.0.14888.3.0.12)).

BIP Schnorr is another variant of EC-SDSA that is not compatible per se with the others, but is close enough to factorize the implementation and use a common ground for all of the variants. See https://crypto.stackexchange.com/questions/34863/ec-schnorr-signature-multiple-standard for a thorough discussion around this.

All-in-all, we can consider that SDSA and EC-SDSA are indeed "recognized by a national or international standards body", and their description is quite unambiguous in the ISO14888-3 standard.

Regards,

**Comment by paulidale** [comment]
Definitely worth reopening this.

**Comment by t8m** [comment]
Please note we are looking for community help implementing this issue. Please comment if you are interested in implementing this feature or if you have reasons why this feature is important enough so we should consider implementing it with higher priority than other requested features. The issue will be closed after 3 months otherwise.

**Comment by t8m** [comment]
This issue was closed because it was marked with the help wanted label and there was no interest in implementing this issue or any other activity for more than 2 months.
